src/main/java/io/vertx/ext/asyncsql/impl/VertxExecutionContext.java

```diff
@@ -0,0 +1,63 @@
+package io.vertx.ext.asyncsql.impl;
+
+import io.vertx.core.Context;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import io.vertx.core.logging.Logger;
+import io.vertx.core.logging.LoggerFactory;
+import scala.concurrent.ExecutionContext;
+
+import java.util.Objects;
+
+/**
+ * Execution environment for Scala Future. By default, Scala uses a <em>reasonable default thread pool.</em>. In the
+ * vert.x context, we use the Worker thread pool. If the task is submitted using a Worker thread, the same thread is
+ * used. If not, the task is delegated to a worker thread.
+ *
+ * @author <a href="http://escoffier.me">Clement Escoffier</a>
+ */
+public class VertxExecutionContext implements ExecutionContext {
+
+  public static ExecutionContext create(Vertx vertx) {
+    return new VertxExecutionContext(vertx, (ex) -> LOGGER.error("An exception occurred", ex));
+  }
+
+  public static ExecutionContext create(Vertx vertx,
+                                        Handler<Throwable> handler) {
+    return new VertxExecutionContext(vertx, handler);
+  }
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(VertxExecutionContext.class);
+
+  private final Vertx vertx;
+  private final Handler<Throwable> errorHandler;
+
+  private VertxExecutionContext(Vertx vertx, Handler<Throwable> errorHandler) {
+    Objects.requireNonNull(errorHandler);
+    Objects.requireNonNull(vertx);
+    this.vertx = vertx;
+    this.errorHandler = errorHandler;
+  }
+
+  @Override
+  public void execute(Runnable runnable) {
+    if (Context.isOnWorkerThread()) {
```

Comment:
On 24/08/15 12:47, Clement Escoffier wrote:
>
> In src/main/java/io/vertx/ext/asyncsql/impl/VertxExecutionContext.java 
> <https://github.com/vert-x3/vertx-mysql-postgresql-client/pull/21#discussion_r37742725>:
>
> > +
> > +  private static final Logger LOGGER = LoggerFactory.getLogger(VertxExecutionContext.class);
> > +
> > +  private final Vertx vertx;
> > +  private final Handler<Throwable> errorHandler;
> > +
> > +  private VertxExecutionContext(Vertx vertx, Handler<Throwable> errorHandler) {
> > +    Objects.requireNonNull(errorHandler);
> > +    Objects.requireNonNull(vertx);
> > +    this.vertx = vertx;
> > +    this.errorHandler = errorHandler;
> > +  }
> > +
> > +  @Override
> > +  public void execute(Runnable runnable) {
> > +    if (Context.isOnWorkerThread()) {
>
> So, it can be replaced by:
>
> |   @Override
>    public void execute(Runnable runnable) {
>      vertx.runOnContext(v -> runnable.run());
>    }|

Nope that won't work.

You need to do context.runOnContext and keep a reference to the context 
in the execution context.

BTW, I answered an almost identical question on the google group today ;)

and also in mod-lang-scala 
https://github.com/vert-x/mod-lang-scala/issues/148#issuecomment-134074097

> |
> |
>
> —
> Reply to this email directly or view it on GitHub 
> <https://github.com/vert-x3/vertx-mysql-postgresql-client/pull/21/files#r37742725>.
>